[docs] Convert serve/llm/quick-start.rst to MyST Markdown (#57782)
ray-project/ray: doc/source/serve/llm/quick-start.md

::::{tab-set}

:::{tab-item} Python
:sync: builder

```python
from ray import serve
from ray.serve.llm import LLMConfig, build_openai_app

llm_config = LLMConfig(
    model_loading_config=dict(
        model_id="qwen-0.5b",
        model_source="Qwen/Qwen2.5-0.5B-Instruct",
    ),
    deployment_config=dict(
        autoscaling_config=dict(
            min_replicas=1, max_replicas=2,
        )
    ),
    log_engine_metrics=False
)

app = build_openai_app({"llm_configs": [llm_config]})
serve.run(app, blocking=True)
```
:::

:::{tab-item} YAML
:sync: bind

```yaml
# config.yaml
applications:
- args:
    llm_configs:
        - model_loading_config:
            model_id: qwen-0.5b
            model_source: Qwen/Qwen2.5-0.5B-Instruct
        accelerator_type: A10G
        deployment_config:
            autoscaling_config:
                min_replicas: 1
                max_replicas: 2
        log_engine_metrics: false
  import_path: ray.serve.llm:build_openai_app
  name: llm_app
  route_prefix: "/"
```
:::

::::

## Advanced usage patterns

> For each usage pattern, we provide a server and client code snippet.

Reviewer: ```suggestion
For each usage pattern, Ray Serve LLM provides a server and client code snippet.
```

        max_loras=16, # Need to set this to the same value as `max_num_adapters_per_replica`.
    ),
    deployment_config=dict(
        autoscaling_config=dict(
            min_replicas=1,
            max_replicas=2,
        )
    ),
    accelerator_type="A10G",
)

# Build and deploy the model
app = build_openai_app({"llm_configs": [llm_config]})
serve.run(app, blocking=True)
```
:::

:::{tab-item} Client
:sync: client

```python
from openai import OpenAI

# Initialize client
client = OpenAI(base_url="http://localhost:8000/v1", api_key="fake-key")

# Make a request to the desired lora checkpoint
response = client.chat.completions.create(
    model="qwen-0.5b:lora_model_1_ckpt",
    messages=[{"role": "user", "content": "Hello!"}],
    stream=True,
)

for chunk in response:
    if chunk.choices[0].delta.content is not None:
        print(chunk.choices[0].delta.content, end="", flush=True)
```
:::

::::

### Embeddings

You can generate embeddings by selecting the embed task in the engine arguments. Models supporting this use case are listed at [vLLM text embedding models](https://docs.vllm.ai/en/stable/models/supported_models.html#text-embedding-task-embed).

> Note: You need to set the `VLLM_USE_V1` environment variable to `0`, since the VLLM V1 doesn't yet fully support the embedding endpoints.

Reviewer: ```suggestion
Note: You need to set the `VLLM_USE_V1` environment variable to `0`, because the VLLM V1 doesn't yet fully support the embedding endpoints.
```

serve.run(app, blocking=True)
```
:::

:::{tab-item} Client (JSON Object)
:sync: client

```python
from openai import OpenAI

# Initialize client
client = OpenAI(base_url="http://localhost:8000/v1", api_key="fake-key")

# Request structured JSON output
response = client.chat.completions.create(
    model="qwen-0.5b",
    response_format={"type": "json_object"},
    messages=[
        {
            "role": "system",
            "content": "You are a helpful assistant that outputs JSON."
        },
        {
            "role": "user",
            "content": "List three colors in JSON format"
        }
    ],
    stream=True,
)

for chunk in response:
    if chunk.choices[0].delta.content is not None:
        print(chunk.choices[0].delta.content, end="", flush=True)
# Example response:
# {
#   "colors": [
#     "red",
#     "blue",
#     "green"
#   ]
# }
```
:::

:::{tab-item} Client (JSON Schema)

If you want, you can also specify the schema you want for the response with pydantic models:

```python
from openai import OpenAI
from typing import List, Literal
from pydantic import BaseModel

# Initialize client
client = OpenAI(base_url="http://localhost:8000/v1", api_key="fake-key")

> # Define a pydantic model of a preset of allowed colors

Reviewer: ```suggestion
# Define a Pydantic model of a preset of allowed colors
```

            max_replicas=2,
        )
    ),
    accelerator_type="L40S",
    engine_kwargs=dict(
        tensor_parallel_size=1,
        max_model_len=8192,
    ),
)

# Build and deploy the model
app = build_openai_app({"llm_configs": [llm_config]})
serve.run(app, blocking=True)
```
:::

:::{tab-item} Client
:sync: client

```python
from openai import OpenAI

# Initialize client
client = OpenAI(base_url="http://localhost:8000/v1", api_key="fake-key")

# Create and send a request with an image
response = client.chat.completions.create(
    model="pixtral-12b",
    messages=[
        {
            "role": "user",
            "content": [
                {
                    "type": "text",
                    "text": "What's in this image?"
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": "https://example.com/image.jpg"
                    }
                }
            ]
        }
    ],
    stream=True,
)

for chunk in response:
    if chunk.choices[0].delta.content is not None:
        print(chunk.choices[0].delta.content, end="", flush=True)
```
:::

::::

> ### Serving small models on fraction of GPUs

Reviewer: ```suggestion
### Serving small models on fractions of GPUs
```

    ),
    engine_kwargs=dict(
        gpu_memory_utilization=0.4,
        use_tqdm_on_load=False,
        enforce_eager=True,
        max_model_len=2048,
    ),
    deployment_config=dict(
        autoscaling_config=dict(
            min_replicas=8, max_replicas=8,
        )
    ),
    accelerator_type="L4",
    placement_group_config=dict(bundles=[dict(GPU=0.49)]),
    runtime_env=dict(
        env_vars={
            "VLLM_RAY_PER_WORKER_GPUS": "0.49",
            "VLLM_DISABLE_COMPILE_CACHE": "1",
        },
    ),
)

app = build_openai_app({"llm_configs": [llm_config]})
serve.run(app, blocking=True)
```

In this example, we've specified 8 replicas of the model, and we use the `placement_group_config` to specify the fraction of the GPU that each model will use in placement group. We also need to set the `VLLM_RAY_PER_WORKER_GPUS` environment variable so that the vLLM GPU workers will each claim a fraction of the GPU. There's also a resource contention [issue](https://github.com/vllm-project/vllm/issues/24601) among workers when doing torch compile caching, so we need to set the `VLLM_DISABLE_COMPILE_CACHE` environment variable to get around it.

> We need to make sure `gpu_memory_utilization` is set to a reasonable value, because vLLM will pre-allocate the GPU memory based on GPU memory utilization, regardless of how we schedule the Ray GPU workers. In this example, setting it to 0.4 means that vLLM will target to use 40% of the GPU memory for the model, its kv-cache, CUDAGraph memory, etc.

Reviewer: ```suggestion
You need to make sure to set `gpu_memory_utilization` to a reasonable value, because vLLM pre-allocates the GPU memory based on GPU memory utilization, regardless of how you schedule the Ray GPU workers. In this example, setting it to 0.4 means that vLLM targets to use 40% of the GPU memory for the model, its kv-cache, CUDAGraph memory, etc.
```

serve.run(app, blocking=True)
```

### How to configure tokenizer pool size so it doesn't hang?

When using `tokenizer_pool_size` in vLLM's `engine_kwargs`, `tokenizer_pool_size` is also required to configure together in order to have the tokenizer group scheduled correctly.

The following example shows a config:

```yaml
# config.yaml
applications:
- args:
    llm_configs:
        - engine_kwargs:
            max_model_len: 1000
            tokenizer_pool_size: 2
            tokenizer_pool_extra_config: "{\"runtime_env\": {}}"
          model_loading_config:
            model_id: Qwen/Qwen2.5-7B-Instruct
  import_path: ray.serve.llm:build_openai_app
  name: llm_app
  route_prefix: "/"
```

## Usage data collection

We collect usage data to improve Ray Serve LLM. We collect data about the following features and attributes:

- Model architecture used for serving.
- Whether JSON mode is used.
- Whether LoRA is used and how many LoRA weights are loaded initially at deployment time.
- Whether autoscaling is used and the min and max replicas setup.
- Tensor parallel size used.
- Initial replicas count.
- GPU type used and number of GPUs used.

> If you'd like to opt-out from usage data collection, you can follow {ref}`Ray usage stats <ref-usage-stats>` to disable it.

Reviewer: ```suggestion
To opt-out of usage data collection, you can follow {ref}`Ray usage stats <ref-usage-stats>` to disable it.
```
